linkerd/protocol/http/src/main/scala/io/buoyant/linkerd/protocol/ZipkinTracePropagator.scala

```diff
@@ -61,28 +59,122 @@ object ZipkinTrace {
   val ZipkinSampleHeader = "x-b3-sampled"
   val ZipkinFlagsHeader = "x-b3-flags"
 
+  /**
+   * The separate "b3" header in b3 single header format :
+   * b3={x-b3-traceid}-{x-b3-spanid}-{if x-b3-flags 'd' else x-b3-sampled}-{x-b3-parentspanid},
+   * where the last two fields are optional.
+   */
+  val ZipkinB3SingleHeader = "b3"
+
+  /**
+   * Handle b3 the same way as finagle handles b3, extract the traceid from "b3" header, write back "X-B3-*" multi
+   * headers, and let the flow continue as if it has received "x-b3-".
+   * @note Copied/adapted from finagle TraceInfo.scala
+   */
+  def convertB3SingleHeaderToMultiHeaders(headers: HeaderMap): Unit = {
+    if (headers.contains(ZipkinB3SingleHeader)) {
+      def handleSampled(headers: HeaderMap, value: String): Unit =
+        value match {
+          case "0" => {
+            headers.set(ZipkinSampleHeader, "0")
+            ()
+          }
+          case "d" => {
+            headers.set(ZipkinFlagsHeader, "1")
+            ()
+          }
+          case "1" => {
+            headers.set(ZipkinSampleHeader, "1")
+            ()
+          }
+          case _ => ()
+        }
+      def handleTraceAndSpanIds(headers: HeaderMap, a: String, b: String): Unit = {
+        headers.set(ZipkinTraceHeader, a)
+        headers.set(ZipkinSpanHeader, b)
+        ()
+      }
+
+      val _ = caseInsensitiveGet(headers, ZipkinB3SingleHeader).map(_.split("-").toList) match {
+        case Some(a) =>
+          a.size match {
+            case 0 =>
+              // bogus
+              ()
+            case 1 =>
+              // either debug flag or sampled
+              handleSampled(headers, a(0))
+            case 2 =>
+              // this is required to be traceId, spanId
+              handleTraceAndSpanIds(headers, a(0), a(1))
+            case 3 =>
+              handleTraceAndSpanIds(headers, a(0), a(1))
+              handleSampled(headers, a(2))
+            case 4 =>
```

Comment:
What happens if the size of the list is more than 4, say 5. I think in this case as this is a non exhaustive match, you will get an exception.  In such cases, just ensure that your match is exhaustive or simply add a default match case

Reply:
Done, added default match, cases where size of list is not as expected will just be ignored.